docs2/site/docs/migrations/migration4.md

```diff
@@ -278,6 +278,149 @@ lock (field)
+{
+    public MoneyType()
+    {
+        Field(x => x.Saldo);
+    }
+}
+```
+
+On the `Field(x => x.Saldo)` line when parsing an expression GraphQL.NET should somehow infer
+that the `Money` CLR type corresponds to the `MoneyType` GraphType. In fact, this cannot be done
+without specifying additional information from the caller. GraphQL.NET can only infer some primitive
+CLR types (`int`, `string`, `DateTime`, `Guid`, etc.) that match built-in scalars.
+
+Type registration is used for the hint:
+
+```csharp
+GraphTypeTypeRegistry.Register<Money, MoneyType>(); // static API before v4
+schema.RegisterTypeMapping<Money, MoneyType>();     // instance method on `ISchema` after v4
+```
+
+Note that since v4 it's possible to register both input and output GraphType for the same CLR type.
+In this case, GraphQL.NET will choose the desired GraphType depending on the context.
+
+```csharp
+public class Money
+{
+    public decimal Amount { get; set; }
+    public string Currency { get; set; }
+}
+
+public class MoneyType : ObjectGraphType<Money>
+{
+    public MoneyType()
+    {
+        Field(x => x.Amount);
+        Field(x => x.Currency);
+    }
+}
+
+public class MoneyInputType : InputObjectGraphType<Money>
+{
+    public MoneyInputType()
+    {
+        Field(x => x.Amount).Description("Total amount").DefaultValue(100m);
+        Field(x => x.Currency).DefaultValue("USD");
+    }
+}
+
+schema.RegisterTypeMapping<Money, MoneyType>();
+schema.RegisterTypeMapping<Money, MoneyInputType>();
+```
+
+An alternative way to define the mapping is to use the new properties in the `GraphQLMetadata` attribute.
+Consider the following example:
+
+```c#
+[GraphQLMetadata(InputType = typeof(FilterInputGraphType))]
+public class Filter
+{
+    public string Key { get; set; }
+    public int Value { get; set; }
+}
+
+public class ContainerRequest
+{
+    public IList<Filter> Filters { get; set; }
+    public int ClientId { get; set; }
+    public int AppId { get; set; }
+}
+
+public class FilterInputGraphType : InputObjectGraphType<Filter>
+{
+    public FilterInputGraphType()
+    {
+        Name = "FilterInput";
+        Field(x => x.Key);
+        Field(x => x.Value);
+    }
+}
+
+public class MyInputType : InputObjectGraphType<ContainerRequest>
+{
+    public MyInputType()
+    {
+        Name = "Input";
+        Field(x => x.Filters); // when building this field, its type is implicitly inferred to list of FilterInputGraphType
+        Field(x => x.ClientId);
+        Field(x => x.AppId, nullable: true);
+    }
+}
+```
+
+In this case, a call to the registration method is not required, since the schema
+will use information from the provided attribute.
+
+> Keep in mind that you can register type mappings even for built-in/primitive types if you want to change their behavior:
+> <br/><br/>
+> schema.RegisterTypeMapping<int, MyIntGraphType>()
+> <br/>
+> schema.RegisterTypeMapping<string, MySpecialFormattedStringGraphType>()
```

Comment:
Needs to be properly formatted.

Reply:
how?

Reply:
```suggestion
Keep in mind that you can register type mappings even for built-in/primitive types if you want to change their behavior:

```csharp
schema.RegisterTypeMapping<int, MyIntGraphType>()
schema.RegisterTypeMapping<string, MySpecialFormattedStringGraphType>()
```

(needs closing triple back quotes also)

Or was it supposed to look like a quote?

Reply:
done